src/Symfony/Component/JsonCrawler/JsonCrawler.php

```diff
@@ -0,0 +1,401 @@
+        'value' => true,
+    ];
+
+    private mixed $data;
+    private JsonPath $evaluatedPath;
+
+    public function __construct(string $json)
+    {
+        try {
+            $this->data = json_decode($json, true, 512, \JSON_THROW_ON_ERROR);
+        } catch (\JsonException $e) {
+            throw new InvalidInputJsonException($e->getMessage());
+        }
+    }
+
+    /**
+     * @return array<array|string|float|int|bool|null>
+     */
+    public function find(string|JsonPath $query): array
+    {
+        $this->evaluatedPath = \is_string($query) ? new JsonPath($query) : $query;
+
+        return $this->evaluate($this->evaluatedPath);
+    }
+
+    private function evaluate(JsonPath $query): array
+    {
+        try {
+            $tokens = JsonPathTokenizer::tokenize($query);
+            $current = [$this->data];
+
+            foreach ($tokens as $token) {
+                $next = [];
+                foreach ($current as $value) {
+                    $result = $this->evaluateToken($token, $value);
+                    $next = array_merge($next, $result);
+                }
+
+                $current = $next;
+            }
+
+            return $current;
+        } catch (InvalidJsonPathException $e) {
+            throw $e;
+        } catch (\Throwable $e) {
+            throw new JsonCrawlerException($query, $e->getMessage());
+        }
+    }
+
+    private function evaluateToken(JsonPathToken $token, mixed $value): array
+    {
+        return match ($token->type) {
+            TokenType::Name => $this->evaluateName($token->value, $value),
+            TokenType::Bracket => $this->evaluateBracket($token->value, $value),
+            TokenType::Recursive => $this->evaluateRecursive($value),
+        };
+    }
+
+    private function evaluateName(string $name, mixed $value): array
+    {
+        if (!\is_array($value)) {
+            return [];
+        }
+
+        if ('*' === $name) {
+            return array_values($value);
+        }
+
+        return \array_key_exists($name, $value) ? [$value[$name]] : [];
+    }
+
+    private function evaluateBracket(string $expr, mixed $value): array
+    {
+        if (!\is_array($value)) {
+            return [];
+        }
+
+        if ('*' === $expr) {
+            return array_values($value);
+        }
+
+        // single negative index
+        if (preg_match('/^-\d+$/', $expr)) {
+            if (!array_is_list($value)) {
+                return [];
+            }
+
+            $index = \count($value) + (int) $expr;
+
+            return isset($value[$index]) ? [$value[$index]] : [];
+        }
+
+        // start and end index
+        if (preg_match('/^-?\d+(?:\s*,\s*-?\d+)*$/', $expr)) {
+            $indices = array_map('trim', explode(',', $expr));
+            if (!array_is_list($value)) {
+                return [];
+            }
+            $result = [];
+            foreach ($indices as $index) {
+                $index = (int) $index;
```

Comment:
The code can be simplified by removing `$indices`.
```suggestion
            foreach (explode(',', $expr) as $index) {
                $index = (int) trim($index);
```

Reply:
Indeed, thanks!